src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+                        'is set properly.'
+                    ].join(' ')));
+                } else {
+                    return reject(new Error([
+                        'unexpected error while fetching topojson file at',
+                        topojsonPath
+                    ].join(' ')));
+                }
+            }
+            resolve(topojson);
+        });
+    });
+};
+
+proto.update = function(geoCalcData, fullLayout) {
+    var geoLayout = fullLayout[this.id];
+
+    // important: maps with choropleth traces have a different layer order
+    this.hasChoropleth = false;
+    for(var i = 0; i < geoCalcData.length; i++) {
+        if(geoCalcData[i][0].trace.type === 'choropleth') {
+            this.hasChoropleth = true;
+            break;
+        }
+    }
+
+    this.updateProjection(fullLayout, geoLayout);
+    this.updateBaseLayers(fullLayout, geoLayout);
+    this.updateDims(fullLayout, geoLayout);
+
+    Plots.generalUpdatePerTraceModule(this, geoCalcData, geoLayout);
+
+    var scatterLayer = this.layers.frontplot.select('.scatterlayer');
+    this.dataPoints.point = scatterLayer.selectAll('.point');
+    this.dataPoints.text = scatterLayer.selectAll('text');
+    this.dataPaths.line = scatterLayer.selectAll('.js-line');
+
+    var choroplethLayer = this.layers.backplot.select('.choroplethlayer');
+    this.dataPaths.choropleth = choroplethLayer.selectAll('path');
+
+    this.updateFx(fullLayout, geoLayout);
+    this.render();
+};
+
+proto.updateProjection = function(fullLayout, geoLayout) {
+    var gs = fullLayout._size;
+    var domain = geoLayout.domain;
+    var projLayout = geoLayout.projection;
+    var rotation = projLayout.rotation || {};
+    var center = geoLayout.center || {};
+
+    var projection = this.projection = getProjection(geoLayout);
+
+    // set 'pre-fit' projection
+    projection
+        .center([center.lon - rotation.lon, center.lat - rotation.lat])
+        .rotate([-rotation.lon, -rotation.lat, rotation.roll])
+        .parallels(projLayout.parallels);
+
+    // setup subplot extent [[x0,y0], [x1,y1]]
+    var extent = [[
+        gs.l + gs.w * domain.x[0],
+        gs.t + gs.h * (1 - domain.y[1])
+    ], [
+        gs.l + gs.w * domain.x[1],
+        gs.t + gs.h * (1 - domain.y[0])
+    ]];
+
+    var lonaxis = geoLayout.lonaxis;
+    var lataxis = geoLayout.lataxis;
+    var rangeBox = makeRangeBox(lonaxis.range, lataxis.range);
+
+    // fit projection 'scale' and 'translate' to set lon/lat ranges
+    projection.fitExtent(extent, rangeBox);
+
+    var b = this.bounds = projection.getBounds(rangeBox);
+    var s = this.fitScale = projection.scale();
+    var t = projection.translate();
+
+    if(
+        !isFinite(b[0][0]) || !isFinite(b[0][1]) ||
+        !isFinite(b[1][0]) || !isFinite(b[1][1]) ||
+        isNaN(t[0]) || isNaN(t[0])
+    ) {
+        Lib.warn('Invalid geo settings');
+
+        // TODO fallback to default ???
```

Comment:
Yes, I would think so - this is the equivalent of setting invalid range on a cartesian plot, right? In that case we autorange and push the autorange back to the input `layout`.

Reply:
Done in https://github.com/plotly/plotly.js/commit/46d29d5af1daedd639742228a746fce1a804b3ac using a `relayout` call for maximum robustness.